File: 3583-count-special-triplets/3583-count-special-triplets.cpp
class Solution {
public:
    int m = 1e9+7;
    int specialTriplets(vector<int>& nums) {
        long long count=0;
        int n = nums.size();
        unordered_map<int,int>right;
        unordered_map<int,int>left;
        for(int num: nums){
            right[num]++;
        }

        for(int num : nums){
            right[num]--;
            if(right[num]==0) right.erase(num);
            if(right.count(num*2) && left.count(num*2)) count= (count+1LL*left[num*2]*right[num*2])%m;
            left[num]++;
        }
        
        return count;
    }
};

File: 3583-count-special-triplets/3583-count-special-triplets.java
class Solution {
    int mod = (int)1e9+7;
    public int specialTriplets(int[] nums) {
        long count=0;
        HashMap<Integer, Integer>right = new HashMap<>();
        HashMap<Integer, Integer>left = new HashMap<>();
        for (int num : nums) {
            right.put(num, right.getOrDefault(num, 0) + 1);
        }

        for(int num: nums){
            right.put(num, right.get(num) - 1);
            if(right.get(num)==0) right.remove(num);
            int target = num*2;
            if(left.containsKey(target) && right.containsKey(target)){
                long ways = (long)left.get(target) * right.get(target);
                count = (count + ways)%mod;
            }
            left.put(num, left.getOrDefault(num, 0) + 1);
        }
        return (int)count;
    }
}